src/transformers/generation_contrastive_search.py

```diff
@@ -0,0 +1,196 @@
+            It saves the cached key-value results that computed by previous steps.
+        last_hidden_states (`torch.FloatTensor`, *optional*):
+            The last_hidden_states generated by the previous step.
+        logit_for_next_step (`torch.FloatTensor`, *optional*):
+            The logit_for_next_step generated by the previous step.
+        is_encoder_decoder (`bool`, *optional*, defaults to `False`):
+            If `True`, the model is an encoder-decode model, otherwise the model is a decoder-only model.
+
+    From: https://github.com/yxuansu/SimCTG/blob/main/contrastive_search_explanation/README.md
+    """
+    bsz, seqlen, embed_dim = last_hidden_states.size()
+    next_probs = nn.functional.softmax(logit_for_next_step, dim=-1)
+    _, top_k_ids = torch.topk(logit_for_next_step, dim=-1, k=beam_width)
+    top_k_probs = torch.gather(next_probs, dim=1, index=top_k_ids)
+    past_key_values = enlarge_past_key_values(past_key_values, beam_width)
+
+    # build next attention mask
+    if "attention_mask" in model_inputs:
+        attention_mask = model_inputs["attention_mask"]  # [B, S]
+        # decoder-only model need the full attention mask, not only the mask for the last token
+        if is_encoder_decoder is False:
+            attention_mask = torch.cat([attention_mask, attention_mask.new_ones((bsz, 1))], dim=-1)
+        attention_mask = attention_mask.unsqueeze(1).expand(-1, beam_width, -1).reshape(-1, attention_mask.size(-1))
+    else:
+        attention_mask = None
+
+    # encoder-decoder model also contains the `encoder_outputs`
+    if is_encoder_decoder and "encoder_outputs" in model_inputs:
+        encoder_outputs = model_inputs["encoder_outputs"]
+    else:
+        encoder_outputs = None
+    next_model_inputs = model.prepare_inputs_for_generation(
+        top_k_ids.view(-1, 1),
+        past=past_key_values,
+        attention_mask=attention_mask,
+        use_cache=True,
+        encoder_outputs=encoder_outputs,
+    )
+    # compute the candidate tokens by the language model and collects their hidden_states
+    output = model(output_hidden_states=True, **next_model_inputs)
+    past_key_values = output.past_key_values
+    logits = output.logits[:, -1, :]
+    # name is different for encoder-decoder and decoder-only models
+    if is_encoder_decoder:
+        next_hidden = output.decoder_hidden_states[-1]
+        full_hidden_states = output.decoder_hidden_states
+    else:
+        next_hidden = output.hidden_states[-1]
+        full_hidden_states = output.hidden_states
+    context_hidden = (
+        last_hidden_states.unsqueeze(1).expand(-1, beam_width, -1, -1).reshape(bsz * beam_width, seqlen, embed_dim)
+    )
+
+    # compute the degeneratin penalty and re-rank the candidates based on the degeneration penalty and the model confidence
+    # the scores and index of the selected tokens are returned
+    selected_scores, selected_idx = ranking_fast(
```

Comment:
@gante I think we can/should make this a logit processor that exceptionally takes `context_hidden` and `next_hidden` as inputs arguments as well . `top_k_probs` can be computed inside the logit processor, `penalty_alpha` and `beam_width` can be stored inside the logic processor at init